crates/pallet-domains/src/runtime_registry.rs

```diff
@@ -56,21 +63,85 @@ impl Default for DomainRuntimeInfo {
+    let mut initial_storages = vec![(total_issuance_key, StorageData(total_issuance.encode()))];
+    for (account_id, balance) in balances {
+        let account_storage_key = sp_domains::domain_account_storage_key(account_id);
+        let account_info = AccountInfo {
+            nonce: domain_runtime_primitives::Nonce::zero(),
+            consumers: 0,
+            // providers are set to 1 for new accounts
+            providers: 1,
+            sufficients: 0,
+            data: pallet_balances::AccountData {
+                free: balance,
+                ..Default::default()
+            },
+        };
+        initial_storages.push((account_storage_key, StorageData(account_info.encode())))
+    }
+
+    initial_storages
+}
+
 impl<Number, Hash> RuntimeObject<Number, Hash> {
     // Return a complete raw genesis with runtime code and domain id set properly
-    pub fn into_complete_raw_genesis(
+    pub fn into_complete_raw_genesis<T: Config>(
         self,
         domain_id: DomainId,
         domain_runtime_info: DomainRuntimeInfo,
-    ) -> RawGenesis {
+        total_issuance: BalanceOf<T>,
+        initial_balances: Vec<(MultiAccountId, BalanceOf<T>)>,
+    ) -> Result<RawGenesis, Error> {
         let RuntimeObject {
             mut raw_genesis, ..
         } = self;
         raw_genesis.set_domain_id(domain_id);
         match domain_runtime_info {
-            DomainRuntimeInfo::EVM { chain_id } => raw_genesis.set_evm_chain_id(chain_id),
+            DomainRuntimeInfo::EVM { chain_id } => {
+                raw_genesis.set_evm_chain_id(chain_id);
+                let initial_balances = initial_balances.into_iter().try_fold(
+                    BTreeMap::<AccountId20, BalanceOf<T>>::new(),
+                    |mut acc, (account_id, balance)| {
+                        let account_id =
+                            domain_runtime_primitives::AccountId20Converter::try_convert_back(
+                                account_id,
+                            )
+                            .ok_or(Error::InvalidAccountIdType)?;
+
+                        let balance = if let Some(previous_balance) = acc.get(&account_id) {
+                            previous_balance
+                                .checked_add(&balance)
+                                .ok_or(Error::BalanceOverflow)?
+                        } else {
+                            balance
+                        };
+
+                        ensure!(
+                            balance >= T::MinInitialDomainAccountBalance::get(),
+                            Error::MinInitialAccountBalance
+                        );
+
+                        acc.insert(account_id, balance);
+                        Ok(acc)
+                    },
+                )?;
+                ensure!(
+                    initial_balances.len() as u32 <= T::MaxInitialDomainAccounts::get(),
+                    Error::MaxInitialDomainAccounts
+                );
```

Comment:
How about moving this check to `can_instantiate_domain` since we are storing the pre-deduplicate `initial_balances` on the runtime state?

Reply:
They do go through the same path but someone can submit a `domain_config` with 1k of entries in `initial_balances` all point to the same account, I believe it is also required by the following benchmark update.